feat(js): Improve JS custom instrumentation docs (#10095)
getsentry/sentry-docs: docs/platforms/javascript/common/performance/instrumentation/custom-instrumentation.mdx

<PlatformContent includePath="enriching-events/import" />

> There are three key APIs for creating spans:

Reviewer: l: Not sure if we call these APIs anywhere else 🤔 

```suggestion
There are three key functions for creating spans:
```

> - [startSpan](#starting-an-active-span-startspan): Creates a new span that is active, and which is automatically ended.

Reviewer: l/q: Given we recommend `startSpan` below - should we already hint this here?

```suggestion
- [startSpan](#starting-an-active-span-startspan): Creates a new span that is active, and which is automatically ended. You'll likely want to use this function.
```

## Active vs. Inactive Spans

When a new span is started, it will automatically be started as a child of the currently active span, if there is one. This means that if a span is started as an **active span**, any spans that are created inside of the callback where the span is active will be children of that span. Additionally, errors will be tied to the currently active span, if there is one.

In contrast, **inactive spans** will never have children automatically associated with them. This is useful if you do not care about capturing child activity.

A key constraint for active spans is that they can only be made active inside of a callback. This constraint exists because otherwise it becomes impossible to associate spans with the correct parent span when working with asynchronous code.

In places where you are not able to wrap executing code in a callback (e.g. when working with hooks or similar) you have to work with inactive spans, and can combine this with [withActiveSpan](#with-active-span) to manually associate child spans with the correct parent span.

## Span Starting Options

> The following options can be used for all span starting methods:

Reviewer: l: I don't particularly care if we go with API/function/method but we should remain consistent

```suggestion
The following options can be used for all span starting functions:
```

Only `name` is required, all other options are optional.

## Starting an Active Span (`startSpan`)

For most scenarios, we recommend to start active spans with `Sentry.startSpan()`. This will start a new span that is active in the provided callback, and will automatically end the span when the callback is done. The callback can be synchronuous, or asynchronous (a promise). In the case of an asynchronous callback, the span will be ended when the promise is resolved or rejected. If the provided callback errors or rejects, the span will be marked as failed.

<PlatformContent includePath="performance/start-span" />

## Starting an Active Span with Manual End (`startSpanManual`)

Sometimes, you do not want the span to be ended automatically when the callback is done. In this case, you can use `Sentry.startSpanManual()`. This will start a new span that is active in the provided callback, but will not automatically end the span when the callback is done. You have to manually end the span by calling `span.end()`.

<PlatformContent includePath="performance/start-span-manual" />

## Starting Inactive Spans (`startInactiveSpan`)

> To add spans that aren't active, you can create independent spans. This is useful for when you have work that is grouped together under a single parent span, but is independent from the current active span. However, in most cases you'll want to create and use the start span API from above.

Reviewer: l:

```suggestion
To add spans that aren't active, you can create independent spans. This is useful when you have work that is grouped together under a single parent span, but is independent from the current active span. However, in most cases you'll want to create and use the [startSpan](#starting-an-active-span-startspan) API from above.
```

## Active vs. Inactive Spans

When a new span is started, it will automatically be started as a child of the currently active span, if there is one. This means that if a span is started as an **active span**, any spans that are created inside of the callback where the span is active will be children of that span. Additionally, errors will be tied to the currently active span, if there is one.

In contrast, **inactive spans** will never have children automatically associated with them. This is useful if you do not care about capturing child activity.

A key constraint for active spans is that they can only be made active inside of a callback. This constraint exists because otherwise it becomes impossible to associate spans with the correct parent span when working with asynchronous code.

In places where you are not able to wrap executing code in a callback (e.g. when working with hooks or similar) you have to work with inactive spans, and can combine this with [withActiveSpan](#with-active-span) to manually associate child spans with the correct parent span.

<PlatformCategorySection supported={['browser']}>

> ## Span Hierarchy in the Browser
> 
> In browser environments, it is not possible to reliably keep track of the active span across asynchronous boundaries. This means that if multiple asynchronous operations are started in parallel, it is not possible to determine which span is the parent of which child span. Imagine the following example:

Reviewer: l/m: I'd suggest addinga TLDR-like intro sentence to this section. Right now, we start with the problem and tell folks about the consequence  4 paragraphs in.

I was thinking of something along these lines. Feel free to adjust as you see fit.

```suggestion
## Span Hierarchy in the Browser

In browser environments, spans are by default collected in a flat hierarchy where every span is the direct child of the root span. You can opt into a more fine grained hierarchy but there are trade-offs. 

The key reason for keeping a flat hierarchy is because in browsers it's impossible to reliably keep track of the active span across asynchronous boundaries. This means that if multiple asynchronous operations are started in parallel, it is not possible to determine which span is the parent of which child span. Imagine the following example:
```

```javascript
Sentry.startSpan({ name: "span 1" }, async () => {
  await fetch("https://example.com/1");
  await fetch("https://example.com/2");
  await fetch("https://example.com/3");
});

Sentry.startSpan({ name: "span 2" }, async () => {
  await fetch("https://example.com/4");
  await fetch("https://example.com/5");
  await fetch("https://example.com/6");
});
```

> In the browser, there would be no way to know that `span 1` is only active inside of its callback, while `span 2` is active in the other callback. Because of this, in realitiy _all_ fetch spans would become children of `span 2`. This is misleading and confusing, which is why by default in the browser, **all spans will become children of the root span** (which is usually the pageload or navigation span). This means that you will always have a flat hierarchy of spans.

Reviewer 2: ```suggestion
In the browser, there would be no way to know that `span 1` is only active inside of its callback, while `span 2` is active in the other callback. Because of this, in realitiy, _all_ fetch spans would become children of `span 2`. This is misleading and confusing, which is why by default in the browser, **all spans will become children of the root span** (which is usually the pageload or navigation span). This means that you will always have a flat hierarchy of spans.
```

This is a tradeoff that we have made to ensure that the data that is captured is accurate and reliable. If you need to capture a more complex hierarchy of spans, you can opt-out of this behavior by setting `parentSpanIsAlwaysRootSpan: false`:

```javascript
Sentry.init({
  parentSpanIsAlwaysRootSpan: false,
});
```

> This will revert to use the full hierarchy behavior, where spans are children of the currently active span. However, this may lead to incorrect data in the case of multiple parallel asynchronous operations - it is up to you to ensure there are not multiple parallel asynchronous operations that start spans in this case.

Reviewer 2: ```suggestion
This will revert to use the full hierarchy behavior, where spans are children of the currently active span. However, this may lead to incorrect data in the case of multiple parallel asynchronous operations - it is up to you to ensure there are no multiple parallel asynchronous operations that start spans in this case.
```

</PlatformCategorySection>

## Span Starting Options

The following options can be used for all span starting functions:

| Option             | Type                        | Description                                                         |
| ------------------ | --------------------------- | ------------------------------------------------------------------- |
| `name`             | `string`                    | The name of the span.                                               |
| `op`               | `string`                    | The operation of the span.                                          |
| `startTime`        | `number`                    | The start time of the span.                                         |
| `attributes`       | `Record<string, Primitive>` | Attributes to attach to the span.                                   |
| `onlyIfParent`     | `boolean`                   | If true, ignore the span if there is no active parent span.         |
| `forceTransaction` | `boolean`                   | If true, ensure this span shows up as transaction in the Sentry UI. |

> Only `name` is required, all other options are optional.

Reviewer 2: l: I think this would be better encoded in the table as another column for each param.

Author: this makes the table become a bit unwieldy, as we don't have a lot of horizontal space 😬

Reviewer 2: Ah, no worries then.

## Starting an Active Span (`startSpan`)

For most scenarios, we recommend to start active spans with `Sentry.startSpan()`. This will start a new span that is active in the provided callback, and will automatically end the span when the callback is done. The callback can be synchronuous, or asynchronous (a promise). In the case of an asynchronous callback, the span will be ended when the promise is resolved or rejected. If the provided callback errors or rejects, the span will be marked as failed.

<PlatformContent includePath="performance/start-span" />

## Starting an Active Span with Manual End (`startSpanManual`)

> Sometimes, you do not want the span to be ended automatically when the callback is done. In this case, you can use `Sentry.startSpanManual()`. This will start a new span that is active in the provided callback, but will not automatically end the span when the callback is done. You have to manually end the span by calling `span.end()`.

Reviewer 2: l: how about having an example here instead of just `Sometimes`?

question: what happens if user's never call `span.end()`?

Author: there is an example below, I guess (with the middleware)?

--> if end is never called, the span is never sent (and also any child spans are never sent)

> Sometimes, you do not want the span to be ended automatically when the callback is done. In this case, you can use `Sentry.startSpanManual()`. This will start a new span that is active in the provided callback, but will not automatically end the span when the callback is done. You have to manually end the span by calling `span.end()`.

Reviewer 2: ```suggestion
Sometimes, you do not want the span to be ended automatically when the callback is done. In this case, you can use `Sentry.startSpanManual()`. This will start a new span that is active in the provided callback, but will not be automatically ended when the callback is done. You have to manually end the span by calling `span.end()`.
```

<PlatformContent includePath="performance/start-span-manual" />

## Starting Inactive Spans (`startInactiveSpan`)

> To add spans that aren't active, you can create independent spans. This is useful when you have work that is grouped together under a single parent span, but is independent from the current active span. However, in most cases you'll want to create and use the [startSpan](#starting-an-active-span-startspan) API from above.

Reviewer 2: ```suggestion
To add spans that aren't active, you can create independent spans. This is useful when you have work that is grouped together under a single parent span, but is independent from the currently active span. However, in most cases you'll want to create and use the [startSpan](#starting-an-active-span-startspan) API from above.
```

## Active vs. Inactive Spans

When a new span is started, it will automatically be started as a child of the currently active span, if there is one. This means that if a span is started as an **active span**, any spans that are created inside of the callback where the span is active will be children of that span. Additionally, errors will be tied to the currently active span, if there is one.

In contrast, **inactive spans** will never have children automatically associated with them. This is useful if you do not care about capturing child activity.

A key constraint for active spans is that they can only be made active inside of a callback. This constraint exists because otherwise it becomes impossible to associate spans with the correct parent span when working with asynchronous code.

In places where you are not able to wrap executing code in a callback (e.g. when working with hooks or similar) you have to work with inactive spans, and can combine this with [withActiveSpan](#with-active-span) to manually associate child spans with the correct parent span.

<PlatformCategorySection supported={['browser']}>

## Span Hierarchy in the Browser

> In browser environments, spans are by default collected in a flat hierarchy where every span is the direct child of the root span. You can opt into a more fine grained hierarchy but there are trade-offs. 

Reviewer 3: ```suggestion
In browser environments, spans are by default collected in a flat hierarchy where every span is the direct child of the root span. You can opt into a more fine-grained hierarchy but there are trade-offs. 
```

The key reason for keeping a flat hierarchy is because in browsers it's impossible to reliably keep track of the active span across asynchronous boundaries. This means that if multiple asynchronous operations are started in parallel, it is not possible to determine which span is the parent of which child span. Imagine the following example:

```javascript
Sentry.startSpan({ name: "span 1" }, async () => {
  await fetch("https://example.com/1");
  await fetch("https://example.com/2");
  await fetch("https://example.com/3");
});

Sentry.startSpan({ name: "span 2" }, async () => {
  await fetch("https://example.com/4");
  await fetch("https://example.com/5");
  await fetch("https://example.com/6");
});
```

> In the browser, there would be no way to know that `span 1` is only active inside of its callback, while `span 2` is active in the other callback. Because of this, in realitiy, _all_ fetch spans would become children of `span 2`. This is misleading and confusing, which is why by default in the browser, **all spans will become children of the root span** (which is usually the pageload or navigation span). This means that you will always have a flat hierarchy of spans.

Reviewer 3: ```suggestion
In the browser, there would be no way to know that `span 1` is only active inside of its callback, while `span 2` is active in the other callback. Because of this, in reality, _all_ fetch spans would become children of `span 2`. This is misleading and confusing, which is why by default in the browser, **all spans will become children of the root span** (which is usually the pageload or navigation span). This means that you will always have a flat hierarchy of spans.
```

## Span Starting Options

The following options can be used for all span starting functions:

| Option             | Type                        | Description                                                         |
| ------------------ | --------------------------- | ------------------------------------------------------------------- |
| `name`             | `string`                    | The name of the span.                                               |
| `op`               | `string`                    | The operation of the span.                                          |
| `startTime`        | `number`                    | The start time of the span.                                         |
| `attributes`       | `Record<string, Primitive>` | Attributes to attach to the span.                                   |
| `onlyIfParent`     | `boolean`                   | If true, ignore the span if there is no active parent span.         |
| `forceTransaction` | `boolean`                   | If true, ensure this span shows up as transaction in the Sentry UI. |

Only `name` is required, all other options are optional.

## Starting an Active Span (`startSpan`)

> For most scenarios, we recommend to start active spans with `Sentry.startSpan()`. This will start a new span that is active in the provided callback, and will automatically end the span when the callback is done. The callback can be synchronuous, or asynchronous (a promise). In the case of an asynchronous callback, the span will be ended when the promise is resolved or rejected. If the provided callback errors or rejects, the span will be marked as failed.

Reviewer 3: ```suggestion
For most scenarios, we recommend to start active spans with `Sentry.startSpan()`. This will start a new span that is active in the provided callback, and will automatically end the span when the callback is done. The callback can be synchronous, or asynchronous (a promise). In the case of an asynchronous callback, the span will be ended when the promise is resolved or rejected. If the provided callback errors or rejects, the span will be marked as failed.
```
